### PR TITLE
AWS S3: Avoid uploading a empty chunk after splitting

### PR DESCRIPTION
**Pull request by mpdn**
Fixes https://github.com/akka/alpakka/issues/2363

### CODE REVIEW

File: s3/src/test/scala/akka/stream/alpakka/s3/impl/DiskBufferSpec.scala [hunk 1]
```diff
@@ -84,4 +84,12 @@ class DiskBufferSpec(_system: ActorSystem)
     }
 
   }
+
+  it should "not emit rest chunk if rest chunk is empty" in {
```
Comment (seglo) [inline s3/src/test/scala/akka/stream/alpakka/s3/impl/DiskBufferSpec.scala:88]: Does this test validate that the extra element is only emitted if the `MemoryBuffer` is used for chunking?

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/MemoryBuffer.scala [hunk 2]
```diff
@@ -42,11 +43,14 @@ import akka.util.ByteString
...
       override def onUpstreamFinish(): Unit = {
-        if (isAvailable(out)) emit()
+        if (isAvailable(out) && (!hasEmitted || buffer.nonEmpty)) emit()
         completeStage()
       }
 
-      private def emit(): Unit = emit(out, Chunk(Source.single(buffer), buffer.size), () => completeStage())
+      private def emit(): Unit = {
+        hasEmitted = true
+        emit(out, Chunk(Source.single(buffer), buffer.size), () => completeStage())
```
Comment (seglo) [inline s3/src/main/scala/akka/stream/alpakka/s3/impl/MemoryBuffer.scala:52]: An alternative impl. would be to empty the buffer as part of the  `andThen` thunk.  Then we don't need `hasEmitted` and just rely on the buffer to signal if there's anything to emit.

**Comment by seglo** [comment]
Can you take a look at the test failure?

https://travis-ci.com/github/akka/alpakka/jobs/364241675#L593

**Comment by mpdn** [comment]
On second thought, I am not really happy with the way I tried to fix this. I'll mark the PR as draft again and change my approach.

**Comment by mpdn** [comment]
There we go. I think this is a better way of solving it instead of changing the behavior of the buffers.

File: s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3IntegrationSpec.scala [hunk 3]
```diff
@@ -443,6 +443,59 @@ trait S3IntegrationSpec
...
+
+  it should "only upload single chunk when exact chunk is followed by an empty ByteString" in {
+    val source: Source[ByteString, Any] = Source[ByteString](
+      chunk :: ByteString.empty :: Nil
+    )
+
+    uploadAndAndCheckParts(source, 1)
+  }
+
+  it should "upload two chunks size of ByteSTrings equals chunk size" in {
```
Comment (ennru) [inline s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3IntegrationSpec.scala:461]:
```suggestion
  it should "upload two chunks size of ByteStrings equals chunk size" in {
```

**Comment by ennru** [comment]
Thank you @mpdn!

**Pull request merged**